libdrgn/python/program.c

```diff
@@ -1040,3 +1037,53 @@ Program *program_from_pid(PyObject *self, PyObject *args, PyObject *kwds)
 	}
 	return prog;
 }
+
+ThreadIterator *Program_threads(Program *self)
+{
+	struct drgn_thread_iterator *iter;
+	struct drgn_error *err = drgn_thread_iterator_create(&self->prog, &iter);
+	if (err)
+		return set_drgn_error(err);
+	ThreadIterator *ret =
+		(ThreadIterator *)ThreadIterator_type.tp_alloc(
+			&ThreadIterator_type, 0);
+	if (!ret) {
+		drgn_thread_iterator_destroy(iter);
+		return NULL;
+	}
+	ret->prog = self;
+	ret->iterator = iter;
+	Py_INCREF(self);
+	return ret;
+}
+
+PyObject *Program_crashed_thread(Program *self)
+{
+	struct drgn_thread *drgn_thread;
+	struct drgn_error *err;
+
+	err = drgn_program_crashed_thread(&self->prog, &drgn_thread);
+	if (err)
+		return set_drgn_error(err);
+
+	return Thread_wrap(drgn_thread, self);
+}
+
+PyObject *Program_thread(Program *self, PyObject *args, PyObject *kwds)
+{
+	static char *keywords[] = {"tid", NULL};
+	struct drgn_error *err;
+	struct index_arg tid = {};
+	struct drgn_thread *drgn_thread;
+
+	if (!PyArg_ParseTupleAndKeywords(args, kwds, "O&:thread", keywords, index_converter,
+					 &tid))
+		return NULL;
+
+	err = drgn_program_find_thread(&self->prog, tid.uvalue, &drgn_thread);
+	if (err)
+		return set_drgn_error(err);
+	if (drgn_thread == NULL)
+		return PyErr_Format(PyExc_LookupError, "thread with tid of %ld not found", tid.uvalue);
+	return Thread_wrap(drgn_thread, self);
```

Comment:
We're leaking the `drgn_thread` here (specifically in the kernel case). You either need to destroy the original `drgn_thread` or add a `drgn_program_find_thread_internal` that initializes a `struct drgn_thread` out parameter rather than returning a malloced one.